File: site/tokenizer/isabelle_tokenizer.js
/**
 * The JavaScript half of §5's tokenizer — the one the Worker runs against a query,
 * where the Python half (`Isabelle_Semantic_Embedding/isabelle_tokenizer.py`) built
 * the index. They must agree byte for byte or the site returns silently wrong
 * results with no error anywhere, so this file is written to be read beside that one:
 * same order, same names, same algorithm, not a second reading of the same prose.
 *
 * Two prohibitions from §5.5, both of which fail silently when broken:
 *
 * - **No table and no character class of its own.** Everything it classifies by comes
 *   from the asset. In particular it never asks JavaScript: `\p{L}` is not
 *   `isalpha()` (145,672 code points against 136,104 under Unicode 15), `\p{Nd}`
 *   rejects `²` where `isdigit()` accepts it, and `\s` disagrees with `isspace()` in
 *   both directions — `\s` takes U+FEFF, `isspace()` takes U+001C to U+001F and
 *   U+0085. All four appear in the corpus (§16.4).
 * - **It refuses an asset whose `tokenizer_rule` it does not implement**, rather than
 *   reading its tables and applying different rules to them.
 *
 * And one hazard peculiar to this side: a **character** is a Unicode code point,
 * never a UTF-16 code unit. 4.15 % of the corpus's expressions carry a character
 * above U+FFFF, so a port that indexes by code unit emits unpaired surrogates that
 * JSON transports intact and no query can ever match. Every loop here iterates code
 * points, and the fold scan works over an array of them rather than over the string.
 */

/** The rules this file implements. An asset built by any other version is refused. */
export const SUPPORTED_TOKENIZER_RULES = new Set([1]);

/**
 * Isabelle's own rule for what names a symbol (`Pure/General/symbol.scala`): `\<`, an
 * optional `^`, a letter, then letters, digits, `_` or `'`, then `>`. Text that does
 * not match is not an escape and reaches token formation as ordinary characters —
 * since D43 removed the `symbol_explode` step this is the only place an escape is
 * recognised at all (§5.1 step 3a). ASCII only, so no Unicode property is involved.
 */
const ESCAPE = /\\<\^?[A-Za-z][A-Za-z0-9_']*>/g;

/** Membership over the inclusive [lo, hi] pairs the asset ships. */
class CodePointSet {
  constructor(ranges) {
    this.bounds = [];
    for (const [lo, hi] of ranges) {
      this.bounds.push(lo, hi + 1);
    }
    // `has` is a parity test over these boundaries, so it needs them non-decreasing —
    // which is to say the ranges ascending and non-overlapping. An asset that breaks
    // that does not fail here without this check: it answers wrongly for every
    // character, silently. The asset is committed and hand-editable, and a
    // hand-written one got it wrong the first time.
    for (let i = 1; i < this.bounds.length; i += 1) {
      if (this.bounds[i] < this.bounds[i - 1]) {
        throw new Error('code-point ranges must be ascending and non-overlapping; got '
                        + JSON.stringify(ranges.slice(0, 8)));
      }
    }
  }

  has(codePoint) {
    let lo = 0;
    let hi = this.bounds.length;
    while (lo < hi) {
      const mid = (lo + hi) >> 1;
      if (this.bounds[mid] <= codePoint) lo = mid + 1;
      else hi = mid;
    }
    return (lo & 1) === 1;
  }
}

const charSet = (s) => new Set(Array.from(s));

export class Tokenizer {
  constructor(asset) {
    const rule = asset.tokenizer_rule;
    if (!SUPPORTED_TOKENIZER_RULES.has(rule)) {
      throw new Error(
        `tokenizer asset declares tokenizer_rule ${JSON.stringify(rule)}, which this ` +
        `implementation does not implement (it implements ` +
        `${[...SUPPORTED_TOKENIZER_RULES].join(', ')}). Refusing to read its tables ` +
        `and apply different rules to them (§5.5).`);
    }
    this.asset = asset;
    // Maps rather than plain objects: an asset is data from a file, and a key such as
    // `__proto__` must be a key and not a way of reaching the prototype.
    this.symbols = new Map(Object.entries(asset.symbols));
    this.fold = new Map(Object.entries(asset.fold));
    // U+21E9, U+21E7, U+2759 — read off the fold table's own keys rather than named
    // here, so a new marker in the table needs no code change.
    this.markers = new Set([...this.fold.keys()].map((k) => Array.from(k)[0]));
    this.letters = new CodePointSet(asset.letters);
    this.digits = new CodePointSet(asset.digits);
    this.spaces = new CodePointSet(asset.spaces);
    this.quasi = charSet(asset.quasi_letters);
    this.discarded = charSet(asset.discarded);
    this.asciiSymbolic = charSet(asset.ascii_symbolic);
    this.rendered = charSet(asset.rendered_subsup);
    this.renderedDigits = charSet(asset.rendered_digits);
    this.separators = charSet(asset.separators);
  }

  // ---- §5.1, steps 1 to 3 --------------------------------------------------

  /** Everything before token formation: NFC, U+007F, and the two symbol passes. */
  normalize(s) {
    const converted = s.normalize('NFC').replaceAll('\u007F', ' ')
      .replace(ESCAPE, (m) => this.symbols.get(m) ?? m);
    return this.foldMarkers(converted);
  }

  /**
   * Step 3b: left to right, two characters at a time, non-overlapping.
   *
   * A marker whose next character is itself a marker consumes it: the pair is not in
   * the fold table, neither character folds, and the second marker cannot begin a
   * pair of its own. So `x⇩1` gives `x₁`, `x⇩⇩1` stays as it is, and `x⇩⇩⇩1` gives
   * `x⇩⇩₁`. Rare enough that the user ruled it not worth fixing, and specified
   * anyway because a port folding each marker separately would diverge here and
   * nowhere else.
   */
  foldMarkers(s) {
    const chars = Array.from(s);
    const out = [];
    let i = 0;
    while (i < chars.length) {
      const pair = i + 1 < chars.length ? chars[i] + chars[i + 1] : null;
      const folded = pair === null ? undefined : this.fold.get(pair);
      if (folded !== undefined) {
        out.push(folded);
        i += 2;
      } else if (pair !== null && this.markers.has(chars[i])) {
        out.push(pair);
        i += 2;
      } else {
        out.push(chars[i]);
        i += 1;
      }
    }
    return out.join('');
  }

  // ---- §5.2 ----------------------------------------------------------------

  tokenize(s) {
    const out = [];
    let cur = [];
    let mode = null;
    const flush = () => {
      if (cur.length) {
        out.push(cur.join(''));
        cur = [];
      }
      mode = null;
    };

    for (const ch of this.normalize(s)) {
      const cp = ch.codePointAt(0);
      if (this.spaces.has(cp) || this.discarded.has(ch)) {
        flush();
      } else if (this.letters.has(cp)
                 || (mode === 'id' && (this.digits.has(cp) || this.quasi.has(ch)))) {
        // A digit continues an identifier in preference to starting a numeral:
        // `x1` is one token, not two.
        if (mode !== 'id') {
          flush();
          mode = 'id';
        }
        cur.push(ch);
      } else if (this.digits.has(cp) && !this.renderedDigits.has(ch)) {
        // A rendered sub/superscript digit is decoration, not content, so it falls
        // through to *anything else* and §5.4's fallback keeps it.
        if (mode !== 'num') {
          flush();
          mode = 'num';
        }
        cur.push(ch);
      } else if (this.asciiSymbolic.has(ch)) {
        if (mode !== 'sym') {
          flush();
          mode = 'sym';
        }
        cur.push(ch);
      } else {
        flush();
        out.push(ch);
      }
    }
    flush();
    return out;
  }

  // ---- §5.4 ----------------------------------------------------------------

  subtokens(tokens) {
    const out = [];
    for (const t of tokens) {
      // Split at every separator and drop what is empty.
      const parts = [];
      let cur = [];
      for (const ch of t) {
        if (this.separators.has(ch)) {
          if (cur.length) {
            parts.push(cur.join(''));
            cur = [];
          }
        } else {
          cur.push(ch);
        }
      }
      if (cur.length) parts.push(cur.join(''));
      if (parts.length) {
        // One at a time, not `out.push(...parts)`: V8 caps spread arity at about
        // 125,000 arguments, and `parts` is bounded only by the token's length.
        for (const p of parts) out.push(p);
      } else if (t.length && Array.from(t).every((c) => this.rendered.has(c))) {
        // A token made entirely of rendered sub/superscripts is real content —
        // `ᶜᵉ`, `ₚₜᵣ`, `²` — and would otherwise vanish. Keeping any token that
        // splits to nothing instead would rescue `_` too and break the query `_wrt`.
        out.push(t);
      }
    }
    return out;
  }

  /** Subtokens, which under D21 is the only level that is indexed or queried. */
  run(s) {
    return this.subtokens(this.tokenize(s));
  }
}

File: site/tokenizer/isabelle_tokenizer.py
r"""The production tokenizer of §5 — the one the site export and the query path share.

Two properties make this file what it is, and both are stated as prohibitions in
§5.5 because a violation of either is silent:

* **It carries no table and no character class of its own.** Everything it
  classifies by comes from the asset (`tokenizer_asset`), so the JavaScript port
  can read the same file and cannot drift by consulting `\p{L}` where this
  consulted `isalpha`.
* **It refuses an asset whose `tokenizer_rule` it does not implement.** The
  namespace name embeds the asset's digest, so an index and its asset cannot come
  apart; this is the other half — that the *code* agrees with the file.

It differs from `site/prototype/tokenize_prototype.py`, which measures rather than
specifies, in exactly two ways: it iterates characters instead of Isabelle symbols
(D43), and it has §5.2's numeric token class, which the prototype lacks entirely.
"""
import bisect
import json
import re
import unicodedata

# The rules this file implements. An asset built by any other version is refused.
SUPPORTED_TOKENIZER_RULES = frozenset({1})

# Isabelle's own rule for what names a symbol (`Pure/General/symbol.scala`): `\<`,
# an optional `^`, a letter, then letters, digits, `_` or `'`, then `>`. Text that
# does not match is not an escape and reaches token formation as ordinary
# characters — since D43 removed the `symbol_explode` step, this is the only place
# an escape is recognised at all (§5.1 step 3a).
_ESCAPE = re.compile(r"\\<\^?[A-Za-z][A-Za-z0-9_']*>")


class _CodePointSet:
    """Membership over the inclusive [lo, hi] pairs the asset ships."""

    def __init__(self, ranges):
        ranges = list(ranges)
        self._bounds = []
        for lo, hi in ranges:
            self._bounds.append(lo)
            self._bounds.append(hi + 1)
        # `__contains__` is a parity test over these boundaries, so it needs them
        # non-decreasing — which is to say the ranges ascending and non-overlapping.
        # An asset that breaks that does not fail here without this check: it answers
        # wrongly for every character, silently. `tokenizer_asset` cannot produce such
        # an asset, but the asset is committed and hand-editable, and a hand-written
        # one got it wrong the first time — `aqb` came back as three tokens.
        if any(b < a for a, b in zip(self._bounds, self._bounds[1:])):
            raise ValueError(
                'code-point ranges must be ascending and non-overlapping; got %r'
                % (ranges[:8],))

    def __contains__(self, ch):
        return bisect.bisect_right(self._bounds, ord(ch)) % 2 == 1


class Tokenizer:
    def __init__(self, asset):
        rule = asset.get('tokenizer_rule')
        if rule not in SUPPORTED_TOKENIZER_RULES:
            raise ValueError(
                "tokenizer asset declares tokenizer_rule %r, which this implementation "
                "does not implement (it implements %s). Refusing to read its tables and "
                "apply different rules to them (\u00a75.5)."
                % (rule, sorted(SUPPORTED_TOKENIZER_RULES)))
        self.asset = asset
        self._symbols = asset['symbols']
        self._fold = asset['fold']
        # `\u21e9`, `\u21e7`, `\u2759` — read off the fold table's own keys rather
        # than named here, so a new marker in the table needs no code change.
        self._markers = frozenset(k[0] for k in self._fold)
        self._letters = _CodePointSet(asset['letters'])
        self._digits = _CodePointSet(asset['digits'])
        self._spaces = _CodePointSet(asset['spaces'])
        self._quasi = frozenset(asset['quasi_letters'])
        self._discarded = frozenset(asset['discarded'])
        self._ascii_symbolic = frozenset(asset['ascii_symbolic'])
        self._rendered = frozenset(asset['rendered_subsup'])
        self._rendered_digits = frozenset(asset['rendered_digits'])
        self._separators = frozenset(asset['separators'])

    # ---- §5.1, steps 1 to 3 -------------------------------------------------

    def normalize(self, s):
        """Everything before token formation: NFC, U+007F, and the two symbol passes."""
        s = unicodedata.normalize('NFC', s).replace('\x7f', ' ')
        s = _ESCAPE.sub(lambda m: self._symbols.get(m.group(0), m.group(0)), s)
        return self._fold_markers(s)

    def _fold_markers(self, s):
        """Step 3b: left to right, two characters at a time, non-overlapping.

        A marker whose next character is itself a marker consumes it: the pair is
        not in the fold table, neither character folds, and the second marker
        cannot begin a pair of its own. So `x⇩1` gives `x₁`, `x⇩⇩1` stays as it is,
        and `x⇩⇩⇩1` gives `x⇩⇩₁`. Rare enough that the user ruled it not worth
        fixing, and specified anyway because a port folding each marker separately
        would diverge here and nowhere else.
        """
        out, i, n = [], 0, len(s)
        while i < n:
            pair = s[i:i + 2]
            folded = self._fold.get(pair) if len(pair) == 2 else None
            if folded is not None:
                out.append(folded)
                i += 2
            elif len(pair) == 2 and pair[0] in self._markers:
                out.append(pair)
                i += 2
            else:
                out.append(s[i])
                i += 1
        return ''.join(out)

    # ---- §5.2 ---------------------------------------------------------------

    def tokenize(self, s):
        out, cur, mode = [], [], None

        def flush():
            nonlocal mode
            if cur:
                out.append(''.join(cur))
                del cur[:]
            mode = None

        for ch in self.normalize(s):
            if ch in self._spaces or ch in self._discarded:
                flush()
            elif (ch in self._letters
                  or (mode == 'id' and (ch in self._digits or ch in self._quasi))):
                # A digit continues an identifier in preference to starting a
                # numeral: `x1` is one token, not two.
                if mode != 'id':
                    flush()
                    mode = 'id'
                cur.append(ch)
            elif ch in self._digits and ch not in self._rendered_digits:
                # A rendered sub/superscript digit is decoration, not content, so
                # it falls through to *anything else* and §5.4's fallback keeps it.
                if mode != 'num':
                    flush()
                    mode = 'num'
                cur.append(ch)
            elif ch in self._ascii_symbolic:
                if mode != 'sym':
                    flush()
                    mode = 'sym'
                cur.append(ch)
            else:
                flush()
                out.append(ch)
        flush()
        return out

    # ---- §5.4 ---------------------------------------------------------------

    def subtokens(self, tokens):
        out = []
        for t in tokens:
            # Split at every separator and drop what is empty. Written out rather
            # than as a character-class regular expression so that the JavaScript
            # port is the same algorithm and not a second reading of one.
            parts, cur = [], []
            for ch in t:
                if ch in self._separators:
                    if cur:
                        parts.append(''.join(cur))
                        del cur[:]
                else:
                    cur.append(ch)
            if cur:
                parts.append(''.join(cur))
            if parts:
                out.extend(parts)
            elif t and all(c in self._rendered for c in t):
                # A token made entirely of rendered sub/superscripts is real
                # content — `ᶜᵉ`, `ₚₜᵣ`, `²` — and would otherwise vanish. Keeping
                # any token that splits to nothing instead would rescue `_` too and
                # break the query `_wrt`.
                out.append(t)
        return out

    def __call__(self, s):
        return self.subtokens(self.tokenize(s))


def load(path):
    with open(path, encoding='utf-8') as f:
        return Tokenizer(json.load(f))
